dmscripts/send_stats_to_performance_platform.py

```diff
@@ -0,0 +1,163 @@
+import json
+
+import backoff
+import base64
+import requests
+
+from datetime import datetime, timedelta
+import dmapiclient
+
+from dmscripts.helpers import logging_helpers
+from dmscripts.helpers.logging_helpers import logging
+
+HOURLY_TIME_FORMAT = '%Y-%m-%dT%H:00:00+00:00'  # On the hour exactly
+DAILY_TIME_FORMAT = '%Y-%m-%dT00:00:00+00:00'  # Midnight
+PERFORMANCE_PLATFORM_URLS = {
+    "day": {
+        "stage": "https://www.performance.service.gov.uk/data/gcloud/applications-by-stage",
+        "lot": "https://www.performance.service.gov.uk/data/gcloud/applications-by-lot",
+    },
+    "hour": {
+        "stage": "https://www.performance.service.gov.uk/data/gcloud/applications-by-stage-realtime",
+        "lot": "https://www.performance.service.gov.uk/data/gcloud/applications-by-lot-realtime",
+    },
+}
+
+logger = logging_helpers.configure_logger({'dmapiclient': logging.WARNING})
+
+
+def _format_statistics(stats, category, groupings):
+    return _label_and_count(stats[category], groupings)
+
+
+def _label_and_count(stats, groupings):
+    data = {
+        label: _sum_counts(stats, filters)
+        for label, filters in groupings.items()
+    }
+    return data
+
+
+def _sum_counts(stats, filter_by=None, sum_by='count'):
+    return sum(
+        statistic[sum_by] for statistic in stats
+        if not filter_by or all(
+            _find(statistic.get(key), value)
+            for key, value in filter_by.items()
+        )
+    )
+
+
+def _find(statistic, filter_value):
+    if isinstance(filter_value, list):
+        return statistic in filter_value
+    else:
+        return statistic == filter_value
+
+
+def _generate_id(timestamp, period, data_type, data_item):
+    # Instructions from Performance Platform:
+    # _id should be a unique url-friendly, base64-encoded, UTF8 encoded concatenation identifier, formed from:
+    # _timestamp, service (= gcloud), period (= day or hour), dataType (= applications-by-stage/lot), stage/lot
+    id_bytes = '{}-gcloud-{}-{}-{}'.format(timestamp, period, data_type, data_item).encode('utf-8')
+    return base64.b64encode(id_bytes)
+
+
+def send_data(data, url, pp_bearer):
+    # Equivalent to
+    # curl -X POST -d '<payload>' -H 'Content-type: application/json' -H 'Authorization: Bearer <bearer-token>' <url>
+    logger.info(u"Sending data to Performance Platform dataset '{url}':\n{data}", extra={'url': url, 'data': data})
+    res = requests.post(url, json=data, headers={'Authorization': 'Bearer {}'.format(pp_bearer)})
+    if res.status_code != 200:
+        logger.error(
+            u"Failed to send data: {code}: {cause}",
+            extra={'code': res.status_code, 'cause': res.json().get('message', res.text)}
+        )
+    return res.status_code
+
+
+def applications_by_stage(stats):
+    return _format_statistics(
+        stats,
+        'interested_suppliers',
+        {
+            'interested': {
```

Comment:
It might be nice to document how all these `_format_statistics` et al. functions are supposed to work - what the meaning of these blocks here is - i.e. that lists cause an `in` test, other types lead to an equality test...